### PR TITLE
refactor(utils): add progress visualisation logic

### LINKED ISSUE #116: Add `progress` logging to CLI

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR includes:

> ⚠️ Note:   
> Next to the listed changes quite some adoptions went into test flakeyness 

- adding a new third party package for the progress bar visualisation `multi-progress-bar`
- a new helper in `helper` called `getProgressBar`
- updates to the `executePlugins` function in `core`
- styling helper to abstract color's and style into states
- condition to show progress-bar depending on CLI arg

As the tests where really flakey I reused more of the same config files and added timestamps to the plugin outputs. This did dot fully sole the issue.

Example string:  
`Run Plugins: ████████████████████████████              70% | Executing:  Async Plugin 7
`  
`Run Plugins: ███████████████████████████████████████▉ 100% | Done running plugins`

You can demo the progress bar by running:  
`nx execute cli collect --config=./packages/cli/test/plugin-progress-bar.config.mock.ts`

https://github.com/flowup/quality-metrics-cli/assets/10064416/eb81746c-1312-45eb-a16c-404733e0b053

closes #116

### CODE REVIEW

File: e2e/cli-e2e/tests/collect.spec.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ describe('CLI collect', () => {
   it('should run ESLint plugin and create report.json', async () => {
     const { code, stderr } = await executeProcess({
       command: 'npx',
-      args: ['../../dist/packages/cli', 'collect'],
+      args: ['../../dist/packages/cli', 'collect', '--no-progress'],
```
Comment (matejchalk) [inline e2e/cli-e2e/tests/collect.spec.ts:35]: It's cool that the negated syntax works :slightly_smiling_face:

File: packages/cli/src/lib/commands.ts [hunk 2]
```diff
@@ -9,8 +9,8 @@ export const commands: CommandModule[] = [
     ...yargsAutorunCommandObject(),
     command: '*',
   },
-  yargsAutorunCommandObject(),
+  yargsConfigCommandObject(),
   yargsCollectCommandObject(),
   yargsUploadCommandObject(),
-  yargsConfigCommandObject(),
+  yargsAutorunCommandObject(),
```
Comment (matejchalk) [inline packages/cli/src/lib/commands.ts:15]: If this is about the help output, then I stick to the original ordering. The `autorun` command is the main command for most users (possibly the only command they will use), whereas `print-config` is more of a helper command for debugging (probably more useful to us than to our users).

File: packages/cli/src/lib/implementation/config-middleware.spec.ts [hunk 3]
```diff
@@ -6,38 +6,27 @@ import { configMiddleware } from './config-middleware';
...
-      '..',
-      '..',
-      '..',
-      'models',
-      'test',
-      'fixtures',
-      'code-pushup.config.mock.',
-    ),
-  )}${ext}`;
+  `${withDirName('../../../test/js-format.config.mock.')}${ext}`;
```
Comment (matejchalk) [inline packages/cli/src/lib/implementation/config-middleware.spec.ts:9]: Always use `path.join`, please :pray: 

```suggestion
  withDirName(
    join(
      '..',
      '..',
      '..',
      'test',
      `js-format.config.mock.${ext}`
    ),
  );
```

File: packages/cli/test/js-format.config.mock.cjs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+              {
+                title: 'dummy-title',
+                slug: 'command-object-audit-slug',
+                value: 0,
+                score: 0,
+              },
+            ],
+          )}')`,
+        ],
+        outputFile: `${outputDir}/out.json`,
```
Comment (matejchalk) [inline packages/cli/test/js-format.config.mock.cjs:42]: Please use `echo` instead of `node -e` and `path.join` instead of hard-coded `/` separator. See changes in #131.

(Same goes for all these `js-format.config.mock.*` files.)

File: packages/core/test/plugin-progress-bar.config.mock.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+              slug: auditSlug(pId, aId),
+              title: auditTitle(aId),
+              value: 0,
+              score: 0,
+            },
+          ],
+        )}'), ${duration});`,
+      ],
+      outputFile,
+    },
```
Comment (matejchalk) [inline packages/core/test/plugin-progress-bar.config.mock.ts:38]: Same as above, use `echo` and `join`, please.

File: packages/models/test/fixtures/eslint-plugin.mock.ts [hunk 6]
```diff
@@ -31,7 +31,7 @@ export function eslintPluginConfig(outputDir = 'tmp'): PluginConfig {
     ...eslintMeta,
     runner: runnerConfig(
       Object.values(ESLINT_AUDITS_MAP),
-      join(outputDir, 'eslint-out.json'),
+      join(outputDir, `eslint-out.json`),
```
Comment (matejchalk) [inline packages/models/test/fixtures/eslint-plugin.mock.ts:34]: Unnecessary change:

```suggestion
      join(outputDir, 'eslint-out.json'),
```

File: packages/models/test/fixtures/runner.mock.ts [hunk 7]
```diff
@@ -2,7 +2,7 @@ import { AuditReport, RunnerConfig } from '../../src';
 
 export function runnerConfig(
   audits: AuditReport[],
-  outputFile = 'tmp/out.json',
+  outputFile = `tmp/out.${Date.now()}.json`,
```
Comment (matejchalk) [inline packages/models/test/fixtures/runner.mock.ts:5]: `path.join`

File: e2e/cli-e2e/tests/help.spec.ts [hunk 8]
```diff
@@ -1,13 +1,15 @@
 import { executeProcess } from '@code-pushup/utils';
 
 describe('CLI help', () => {
-  it('should print help', async () => {
+  it('should print help whit argument ---help', async () => {
```
Comment (matejchalk) [inline e2e/cli-e2e/tests/help.spec.ts:4]:
```suggestion
  it('should print help with argument ---help', async () => {
```

File: e2e/cli-e2e/tests/help.spec.ts [hunk 9]
```diff
@@ -1,13 +1,15 @@
...
     const { code, stdout, stderr } = await executeProcess({
       command: 'npx',
       args: ['./dist/packages/cli', '--help'],
     });
     expect(code).toBe(0);
     expect(stderr).toBe('');
-    expect(stdout).toMatchSnapshot();
+    expect(stdout).toMatchSnapshot('help');
```
Comment (matejchalk) [inline e2e/cli-e2e/tests/help.spec.ts:11]: Why? You didn't update the snapshots BTW. But I don't see we shouldn't keep the default snapshot name.

image

**Review by matejchalk — CHANGES_REQUESTED** [review]

File: e2e/cli-e2e/tests/collect.spec.ts (hunk 1 above)
Comment (BioPhoton) [reply e2e/cli-e2e/tests/collect.spec.ts:35]: We picked the right lib at the beginning it seems. I was also impressed that the "dot-syntax" just works

File: packages/cli/test/code-pushup.config.ts [hunk 10]
```diff
@@ -19,29 +28,26 @@ export default {
...
-          label: '???',
           docsUrl: 'http://www.my-docs.dev?slug=dummy-audit-3',
         },
       ],
       runner: {
         command: 'node',
         args: [
-          '-e',
-          `require('fs').writeFileSync('${outputDir}/dummy-plugin-output.json', '${JSON.stringify(
+          'echo',
```
Comment (IKatsuba) [inline packages/cli/test/code-pushup.config.ts:49]: I guess `echo` must be like a command, not an argument. The example here #131

File: packages/core/src/lib/implementation/execute-plugin.ts [hunk 11]
```diff
@@ -105,18 +112,33 @@ export async function executePlugin(
...
+    void 0;
+  },
+};
+
 /**
  * Execute multiple plugins and aggregates their output.
  * @public
- * @param plugins array of {@link PluginConfig} objects
+ * @param plugins array of {@param options
+@link PluginConfig} objects
```
Comment (IKatsuba) [inline packages/core/src/lib/implementation/execute-plugin.ts:141]: Probably, there is a typo.

```suggestion
 * @param plugins array of {@link PluginConfig} objects
 * @param {Object} [options] execution options
 * @param {boolean} options.progress show progress bar
```

File: packages/models/test/fixtures/eslint-plugin.mock.ts (hunk 6 above)
Comment (IKatsuba) [reply packages/models/test/fixtures/eslint-plugin.mock.ts]: The project can enable the eslint rule [quotes](https://typescript-eslint.io/rules/quotes/) to reduce unnecessary template strings.

File: packages/cli/src/lib/implementation/config-middleware.spec.ts (hunk 3 above)
Comment (IKatsuba) [reply packages/cli/src/lib/implementation/config-middleware.spec.ts]: Maybe create a simple `normalize` function
```ts
function normalize(path: string): string {
  return join(path.split('/'));
}
```

`join` looks bulky here.

File: packages/cli/src/lib/commands.ts (hunk 2 above)
Comment (BioPhoton) [reply packages/cli/src/lib/commands.ts]: Ok makes more sense to take the user perspective!

File: packages/cli/test/code-pushup.config.ts (hunk 10 above)
Comment (BioPhoton) [reply packages/cli/test/code-pushup.config.ts:51]: Yes, will change it!

File: packages/core/src/lib/implementation/execute-plugin.ts (hunk 11 above)
Comment (BioPhoton) [reply packages/core/src/lib/implementation/execute-plugin.ts]: adressed

File: packages/cli/src/lib/implementation/config-middleware.spec.ts [hunk 12]
```diff
@@ -48,12 +43,12 @@ describe('applyConfigMiddleware', () => {
...
-  it('should provide default config', async () => {
+  it('should provide default config path', async () => {
     const defaultConfigPath = 'code-pushup.config.js';
     let error: Error = new Error();
     await configMiddleware({ config: defaultConfigPath }).catch(
       e => (error = e),
     );
     expect(error?.message).toContain(defaultConfigPath);
-  });
+  }); /**/
```
Comment (matejchalk) [inline packages/cli/src/lib/implementation/config-middleware.spec.ts:53]:
```suggestion
  });
```

File: packages/core/src/lib/implementation/execute-plugin.ts [hunk 13]
```diff
@@ -105,18 +114,34 @@ export async function executePlugin(
...
+  },
+  updateTitle: (_: string) => {
+    _;
+    void 0;
+  },
+  endProgress: (_?: string) => {
+    _;
+    void 0;
+  },
+};
```
Comment (matejchalk) [inline packages/core/src/lib/implementation/execute-plugin.ts:137]: This confused me, it seems like it shouldn't be in production code, because it's named `MOCK_` and has a very "stubby" implementation. Maybe the intent would be a bit more clear like this?

```suggestion
const NOOP_PROGRESS: ProgressBar = {
  incrementInSteps: () => { /* do nothing */ },
  updateTitle: () => { /* do nothing */ },
  endProgress:() => { /* do nothing */ },
};
```

File: packages/core/src/lib/implementation/execute-plugin.ts [hunk 14]
```diff
@@ -134,10 +159,25 @@ export async function executePlugin(
...
+
+    progressBar.updateTitle(`Executing  ${chalk.bold(pluginCfg.title)}`);
+
     const pluginReport = await executePlugin(pluginCfg);
+    progressBar.incrementInSteps(plugins.length);
+
     return outputs.concat(pluginReport);
   }, Promise.resolve([] as PluginReport[]));
+
+  progressBar.endProgress('Done running plugins');
```
Comment (matejchalk) [inline packages/core/src/lib/implementation/execute-plugin.ts:180]: An alternative to my previous comment - we don't necessarily need to stubbed progress bar at all if we use optional chaining in the method calls :thinking: 

```suggestion
  const progressBar = progress ? getProgressBar(progressName) : null;

  const pluginsResult = await plugins.reduce(async (acc, pluginCfg) => {
    const outputs = await acc;

    progressBar?.updateTitle(`Executing  ${chalk.bold(pluginCfg.title)}`);

    const pluginReport = await executePlugin(pluginCfg);
    progressBar?.incrementInSteps(plugins.length);

    return outputs.concat(pluginReport);
  }, Promise.resolve([] as PluginReport[]));

  progressBar?.endProgress('Done running plugins');
```

File: packages/core/test/plugin-progress-bar.config.mock.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  return {
+    slug: pluginSlug(pId),
+    title: pluginTitle(pId),
+    icon: 'javascript',
+    audits: [{ slug: auditSlug(pId, aId), title: auditTitle(aId) }],
+    runner: {
+      command: 'node',
+      args: [
+        '-e',
+        `setTimeout(() => require('fs').writeFileSync('${outputFile}', '${JSON.stringify(
```
Comment (matejchalk) [inline packages/core/test/plugin-progress-bar.config.mock.ts:26]: As discussed on Slack, this is likely to fail on Windows shells, so we should test it works on Windows first before merging this (hopefully will be part of CI checks soon once #166 is merged).

The simplest workaround I can think of is to invoke the runner as a separate Node script instead of inlining the code (sidesteps all the complexity of how different shells handle quotes and split commands).

File: packages/models/README.md [hunk 16]
```diff
@@ -24,7 +24,7 @@ pnpm add --save-dev @code-pushup/models
 
 Import the type definitions if using TypeScript:
 
-- in `code-pushup.config.ts`:
+- in `cli-parsing.config.ts`:
```
Comment (matejchalk) [inline packages/models/README.md:27]: Should be reverted:

```suggestion
- in `code-pushup.config.ts`:
```

File: packages/models/test/fixtures/report.mock.ts [hunk 17]
```diff
@@ -6,7 +6,7 @@ import { lighthousePluginReport } from './lighthouse-plugin.mock';
 export function report(): Report {
   return {
     packageName: '@code-pushup/core',
-    version: '0.1.0',
+    version: '0.0.1',
```
Comment (matejchalk) [inline packages/models/test/fixtures/report.mock.ts:9]: Why change it? :raised_eyebrow:

File: packages/models/test/fixtures/runner.mock.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+export function runnerConfig(
+  audits: AuditReport[],
+  outputFile = join('tmp', `out.${Date.now()}.json`),
+): RunnerConfig {
+  return {
+    command: 'echo',
+    args: [`${JSON.stringify(audits)} > ${outputFile}`],
+    outputFile,
+  };
+}
```
Comment (matejchalk) [inline packages/models/test/fixtures/runner.mock.ts:13]: I don't think this is being used, is it? Looks like a duplicate of `echoRunnerConfig` without the platform-dependant JSON handling.

The default `outputFile` parameter of `join('tmp', `out.${Date.now()}.json`)` could be useful, though :thinking: Maybe you can add it to `echoRunnerConfig` and remove this function?

**Review by matejchalk — CHANGES_REQUESTED** [review]

File: packages/models/test/fixtures/report.mock.ts (hunk 17 above)
Comment (BioPhoton) [reply packages/models/test/fixtures/report.mock.ts:9]: It happened to be different… flakyness?

File: packages/models/test/fixtures/runner.mock.ts (hunk 18 above)
Comment (BioPhoton) [reply packages/models/test/fixtures/runner.mock.ts:13]: @Tlacenka should have moved her echo fn in here of just adopt the code.

I will refactor in in another PR

File: packages/core/src/lib/implementation/execute-plugin.ts (hunk 14 above)
Comment (BioPhoton) [reply packages/core/src/lib/implementation/execute-plugin.ts:171]: I wanted to implement the „Null Object Pattern“ therefore I went with a stub

File: packages/core/test/plugin-progress-bar.config.mock.ts (hunk 15 above)
Comment (BioPhoton) [reply packages/core/test/plugin-progress-bar.config.mock.ts]: done

File: packages/models/test/fixtures/runner.mock.ts (hunk 18 above)
Comment (Tlacenka) [reply packages/models/test/fixtures/runner.mock.ts:13]: This function was replaced by echoRunnerConfig in my Windows PR, as far as I remember.

File: packages/core/test/plugin-progress-bar.config.mock.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    .pop() || '6',
+);
+
+const outputDir = './tmp';
+const pluginProcess = join(
+  'packages',
+  'core',
+  'test',
+  `plugin-progress-bar-plugin-process.mock.mjs`,
+);
```
Comment (matejchalk) [inline packages/core/test/plugin-progress-bar.config.mock.ts:23]: This way it'll be independent of CWD:

```suggestion
const pluginProcess = join(
  fileURLToPath(dirname(import.meta.url)),
  'plugin-progress-bar-plugin-process.mock.mjs',
);
```

File: packages/core/test/plugin-progress-bar.config.mock.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+const pluginSlug = (id: string): string => 'progress-mock-plugin-' + id;
+const auditSlug = (pId: string, aId: string): string =>
+  pluginSlug(pId) + '-a' + aId;
+const pluginTitle = (end: string): string => 'Async Plugin ' + end;
+const auditTitle = (end: string): string => 'Async Audit ' + end;
+const asyncPlugin = (pId: string, duration = 1000) => {
+  const aId = '0';
+  const outputFile = `${outputDir}/${pluginSlug(pId)}-output.json`;
```
Comment (matejchalk) [inline packages/core/test/plugin-progress-bar.config.mock.ts:31]:
```suggestion
  const outputFile = join(outputDir, `${pluginSlug(pId)}-output.json`);
```

File: packages/core/test/plugin-progress-bar.config.mock.ts [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+      args: [
+        pluginProcess,
+        `--verbose`,
+        `--duration=${duration}`,
+        `--steps=${1}`,
+        `--throwError=${0}`,
+        `--outputDir=${outputDir}`,
+        `--pluginPostfix=${pId}`,
+        `--auditPostfix=${aId}`,
+      ],
```
Comment (matejchalk) [inline packages/core/test/plugin-progress-bar.config.mock.ts:48]: Nice :+1:

File: packages/models/test/fixtures/report.mock.ts (hunk 17 above)
Comment (matejchalk) [reply packages/models/test/fixtures/report.mock.ts:9]: This definitely shouldn't be flaky :sweat_smile: You changed the version in `packages/utils/src/lib/__snapshots__/report-to-md.spec.ts.snap` too. If you change them both back, I don't see why there should be an issue. If there is, it's probably something we need to know about.

**Review by matejchalk — CHANGES_REQUESTED** [review]

File: tmp/.gitkeep
Comment (Tlacenka) [inline tmp/.gitkeep:1]: Is this change intentional? the `.gitkeep` keeps getting deleted for me randomly.

File: packages/models/test/fixtures/runner.mock.ts (hunk 18 above)
Comment (Tlacenka) [reply packages/models/test/fixtures/runner.mock.ts:13]: The reason why I created a new name for the function was that `runnerConfig` does not tell me what command or function the configuration serves. I still see no reason to have the duplicate here when the default value could have just been inserted to the `echoRunnerConfig` function.

File: tmp/.gitkeep
Comment (matejchalk) [reply tmp/.gitkeep:1]: Same for me. It shouldn't have been added to the repo, though, so deleting it is correct.

Comment (BioPhoton) [reply tmp/.gitkeep:1]: this will get removed in a follow up PR of #174

**Comment by BioPhoton** [comment]
https://github.com/flowup/quality-metrics-cli/pull/139#discussion_r1375427865 will get fixed in #180

File: packages/models/test/fixtures/runner.mock.ts (hunk 18 above)
Comment (BioPhoton) [reply packages/models/test/fixtures/runner.mock.ts:13]: > This function was replaced by echoRunnerConfig in my Windows PR, as far as I remember.
It got not renamed ... will merge them in 1 file

File: packages/models/test/fixtures/report.mock.ts (hunk 17 above)
Comment (BioPhoton) [reply packages/models/test/fixtures/report.mock.ts:9]: will be resolved with #174 followup and #180

**Review by Tlacenka — COMMENTED** [review]
From what folder should I run the progress bar to get the example you mentioned working? When I run it from the root folder, I get the wrong score.
image
